CRM/Contact/Form/Task/PDFLetterCommon.php

```diff
@@ -442,34 +497,62 @@ public static function createActivities($form, $html_message, $contactIds) {
     if (!empty($form->_activityId)) {
       $activityParams += array('id' => $form->_activityId);
     }
-    if ($form->_cid) {
-      $activity = CRM_Activity_BAO_Activity::create($activityParams);
-      if (!empty($form->_caseId)) {
-        $caseActivityParams = array('activity_id' => $activity->id, 'case_id' => $form->_caseId);
-        CRM_Case_BAO_Case::processCaseActivity($caseActivityParams);
-      }
+
+    // This seems silly, but the old behavior was to first check `_cid`
+    // and then use the provided `$contactIds`. Probably not even necessary,
+    // but difficult to audit.
+    $contactIds = $form->_cid ? array($form->_cid) : $contactIds;
+
+    $activityContacts = CRM_Core_OptionGroup::values('activity_contacts', FALSE, FALSE, FALSE, NULL, 'name');
+
+    $activityIds = array();
+    switch (Civi::settings()->get('recordGeneratedLetters')) {
+      case 'none':
+        return array();
+
+      case 'multiple':
+        // One activity per contact.
+        foreach ($contactIds as $contactId) {
+          $activity = CRM_Activity_BAO_Activity::create($activityParams);
+          $activityIds[$contactId] = $activity->id;
+          $activityTargetParams = array(
```

Comment:
The code block 

```
          $activityTargetParams = array(
            'activity_id' => $activity->id,
            'contact_id' => $contactId,
            'record_type_id' => CRM_Utils_Array::key('Activity Targets', $activityContacts),
          );
          CRM_Activity_BAO_ActivityContact::create($activityTargetParams);
```

is duplicated in the `case 'combined':` and `case 'combined-attached`, do you think it could be moves into a function?